### PR TITLE
perf(Gemma4): reduce comms in sliding window, FA implementation,

### PR DESCRIPTION
**Pull request by nathan-az**
# What does this PR do ?

Accelerate training for Gemma 4 with Automodel for long-context training with context parallelism

# Changelog

- Reduces the amount collected during CP for sliding window attention to only the required ranks
- Adds FA as an alternative to flex for sliding window layers
- Corrects a cutedsl dependency I encountered on H200s

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Automodel/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?

If you haven't finished some of the above items you can still open "Draft" PR.

# Additional Information

I compared this againts a modified Bridge implementation I have which uses a KV allgather for global layers (ffpa for attn) and TE's attention for sliding window layers. This was about 30% faster. Unclear if this is due to using TE's attention directly which is CP-aware but expects zigzag layout rather than Automodel's contiguous, or if the kernel used is faster (FA3?). This still closes the gap quite significantly. If further optimisation is needed, changing the CP layout unlocks this but is a lot of effort.

Draft was heavily written by Opus – needs proper review. Local equivalence tests on sliding window look healthy (i.e. testing CP8 flash against CP1 with seeded data and grad checking).

| # | Sliding ring    | Backend | Tokens/sec | Peak VRAM |
|---|-----------------|---------|------------|-----------|
| 1 | full (baseline) | flex    |     4,194  |  84.3 GB  |
| 2 | window-aware    | flex    |     5,638  |  80.5 GB  |
| 3 | full (baseline) | flash   |     5,548  |  94.6 GB  |
| 4 | window-aware    | flash   |     5,992  |  82.7 GB  |

Above results were tested with CP8 on 131k context length on 8xH200s.

- Related to # (issue)

### CODE REVIEW

**Comment by Butterfingrz** [comment]
@nathan-az  @akoumpa  Hi, I confirmed that the CuTeDSL 4.6.0 compatibility issue is caused by three Python/NVVM binding API changes: the `ThrMma` type change, and the updated function signatures of `nvvm.fmax` and `nvvm.atomicrmw`.

I will fix this in ffpa-attn to make it compatible with the latest CuTeDSL 4.6.0.

**Comment by nathan-az** [comment]
@Butterfingrz I think it's [from FFPA](https://github.com/xlite-dev/ffpa-attn/blob/882989c2a3aa465867de9ea2dbd19dbed08660d9/src/ffpa_attn/cute/_fwd_d512_sm90.py#L35), specifically

```
from cutlass.cute.nvgpu import OperandMajorMode, cpasync, warpgroup
```
OperandMajorGroup is [not present <4.5](https://github.com/NVIDIA/cutlass/blob/b18df7d206c53394bcdbd71dca917fed0842e684/python/CuTeDSL/cutlass/cute/nvgpu/common.py#L25-L35), but [is present in 4.5.*.](https://github.com/NVIDIA/cutlass/blob/4552152794e8bd3bcfd63cf9b44369e590420dba/python/CuTeDSL/cutlass/cute/nvgpu/common.py#L26-L29)

I'm not sure how to check 4.6.* (looks like it's not being shipped directly from the cutlass repo, or at least doesn't have a tag) so I'm not sure why I had issues with that, but I suspect it was still related to the `cutlass.cute.nvgpu` path.

**Comment by akoumpa** [comment]
Thanks @Butterfingrz that would be awesome , when ready please feel free to ping us to open a PR to upgrade the package -- version bumping will need one AM developer because otherwise the automatic uv lock generation process with fail.

Thanks again!

**Comment by Butterfingrz** [comment]
@akoumpa  Sure! I’ll get this fixed and run the tests as soon as possible.

**Comment by nathan-az** [comment]
@akoumpa any unit tests (or other tests) that should be added to ensure this has no issues? Are there good examples I can use as a template?

And regarding the lockfile to fix the CI – are we expecting that my dependency change will be removable based on FPAA or should I update the lockfile too? (CC @Butterfingrz on this one)

**Comment by akoumpa** [comment]
Hi @nathan-az , first of all, thank you so much for the PR!

Now, rgd testing, i would add a functional test that exercises the specific behavior of the modules this PR modifies, so as to ensure that there's no regression in the future.

For the lockfile, let me open a PR to update the pyproject and then you can rebase and work-around the issue -- thank you for your understanding. Please see here https://github.com/NVIDIA-NeMo/Automodel/pull/3056

**Comment by Butterfingrz** [comment]
@akoumpa @nathan-az Hi! The next release of ffpa-attn will add training support for the SM100 architecture. Please give me a little more time to complete the development.

I also noticed that AutoModel is integrating libraries such as Quack. Is the project now standardizing on NVIDIA CUTLASS DSL 4.6.0.

**Comment by nathan-az** [comment]
Amazing! @akoumpa let me know how the PR is looking. I've upgraded from draft. Still majority vibecoded but the testing suite now looks comprehensive. If you guys have a good test task to actually run (i.e. SFT on X dataset, eval on Y benchmark) I'm happy to run it and post results to improve confidence.

**Comment by akoumpa** [comment]
Hi @nathan-az , thank you for the PR! 🙇 

Since this is gemma4, @athitten could you review when you get the chance?

File: nemo_automodel/components/models/gemma4_moe/cp_attention.py [hunk 1]
```diff
@@ -1126,6 +1197,13 @@ def _run_gemma4_cp_ring_attention(attention_module: torch.nn.Module, ctx: Any) -
     """
     if _ring_use_ffpa_varlen(attention_module, ctx):
         return _Gemma4FFPAVarlenRingAttention.apply(ctx.query, ctx.key, ctx.value, ctx)
+    # Sliding-window layers: optional FlashAttention-2 backend off compiled flex, fully
+    # autograd (no analytical backward, no recompute). Only the plain causal+window+packed
+    # case; vision-bidirectional / padding-mask batches stay on flex. The gate is rank-uniform
+    # (backend + sliding_window); choosing "fa" asserts no vision-bidirectional sliding mask.
+    sliding_backend = getattr(attention_module, "_gemma4_cp_sliding_backend", "flex")
+    if sliding_backend == "fa" and getattr(attention_module, "sliding_window", None) is not None:
```
Comment (HuiyingLi) [inline nemo_automodel/components/models/gemma4_moe/cp_attention.py]: Confirmed a correctness issue at this dispatch on PR head `0c1470c2f62478f6f7d31873a118bdfab3bcc600`.

I ran the production `_run_gemma4_cp_ring_attention -> _Gemma4LocalKernelRingAttention` path with real FlashAttention 2.7.4 on H100s. The input was one packed four-token vision group with zero Q/K and V values `[1, 10, 100, 1000]`, making the visibility mask directly observable. Gemma4 visual-bidirectional semantics produce `[277.75, 277.75, 277.75, 277.75]`; this branch selected `LOCAL_FA` and produced `[1.0, 5.5, 37.0, 278.0]`, matching causal-only rather than visual-bidirectional attention (`max_abs_diff=276.75`).

I repeated the test with CP=2 on two H100s, exercising real NCCL ring exchange plus FA2 forward/backward. A seeded-random parity run gave:

```text
       max diff vs causal   max diff vs visual-bidirectional
out          0.00724                    2.90459
dQ           0.00724                    0.67084
dK           0.01135                    0.50377
dV           0.01290                    1.92629
```

The dispatch entered local FA on both ranks. The local FA calls use `causal=True`, so the `same_vision_group` mask applied by the Flex path is bypassed when `use_bidirectional_attention == "vision"` and `cp_sliding_attn_backend == "fa"`.

**Comment by HuiyingLi** [comment]
I ran FFPA loss-correctness validation at both module and end-to-end levels on H100s.

### Module-level loss parity

At module level this is a scalar attention loss, not language-model cross-entropy: FFPA and the reference receive identical seeded inputs and the same scalar objective.

- **Dense FFPA vs PyTorch float32 SDPA math:** **8/8 cases passed**. Coverage included sequence lengths 64, 256, 1024, and 4096; causal and non-causal attention; and MQA/GQA/MHA layouts (Hq=8, Hkv=1/4/8, head dim 512).
- **Real CP8 FFPA ring vs full-sequence float32 SDPA:** **3/3 cases passed**: two single-document cases at sequence length 2048 with different seeds, plus one packed sequence crossing CP shard boundaries at sequence length 1024.
- The repository GPU functional suite also passed **7/7** cases covering single-card and CP attention loss behavior, packed/unpacked sequences, and MHA/GQA.

### End-to-end loss validation

Config: **google/gemma-4-E2B-it**, BF16, identical seed/data/optimizer, 8192 tokens, 5 training steps on H100. I compared FFPA CP8 with a Flex CP8 control and full-sequence PyTorch SDPA CP1.

| Step | SDPA CP1 loss | Flex CP8 loss | FFPA CP8 loss |
|---:|---:|---:|---:|
| 0 | 6.014351 | 6.014396 | 6.015386 |
| 1 | 5.117459 | 5.080273 | 5.080309 |
| 2 | 4.780338 | 4.774234 | 4.768191 |
| 3 | 4.583474 | 4.580564 | 4.579087 |
| 4 | 4.464567 | 4.459867 | 4.457848 |

The FFPA loss curve stayed aligned with both references and decreased at every step:

- FFPA vs SDPA: maximum absolute step loss difference **0.037149** (**0.726%**); final loss difference **0.006719** (**0.150%**).
- FFPA vs Flex: maximum absolute step loss difference **0.006043** (**0.127%**); final loss difference **0.002018** (**0.045%**).

W&B runs:

- [FFPA CP8](https://wandb.ai/Nemo-automodel/pr2984-ffpa-sdpa-e2e/runs/m64iexvs)
- [Flex CP8 control](https://wandb.ai/Nemo-automodel/pr2984-ffpa-sdpa-e2e/runs/5a6ri5po)
- [PyTorch SDPA full-sequence reference](https://wandb.ai/Nemo-automodel/pr2984-ffpa-sdpa-e2e/runs/iutl132p)

The E2E data is text-only (num_images_per_sample=0), so this specifically validates the FFPA training-loss path.

**Comment by HuiyingLi** [comment]
I also ran a dedicated end-to-end performance validation for the PR optimization.

### Setup

- Model: **google/gemma-4-E2B-it**, BF16 training
- Hardware: **8x H100 80GB**, CP8
- Sequence length: **32,768**, global batch size 1
- 35 language layers: 7 global-attention layers and 28 sliding-window layers
- The 7 global layers used FFPA in every mode; the sweep isolates the sliding-window KV-ring reduction and sliding backend.
- 3 warmup steps followed by 12 measured training steps
- Step time is the maximum across ranks; the table reports medians to reduce the effect of periodic training-step spikes.
- Text-only mock data, so this measures the optimized text path rather than the vision fallback.

### Main 12-step result

| Sliding path | Median step time | Median throughput | Speedup vs full-ring Flex | Peak memory |
|---|---:|---:|---:|---:|
| Full KV ring + Flex | 1.491 s | 21.97k tokens/s | 1.000x | 23.18 GiB |
| Reduced KV ring + Flex | 1.163 s | 28.16k tokens/s | **1.282x** | 23.04 GiB |
| Full KV ring + local FA | 1.101 s | 29.76k tokens/s | **1.355x** | 24.31 GiB |
| Reduced KV ring + local FA | 1.055 s | 31.05k tokens/s | **1.413x** | 23.65 GiB |

This separates the effects:

- KV-ring reduction alone, with Flex held fixed: **1.282x**.
- Local FA alone, with the full KV ring held fixed: **1.355x**.
- Combined PR path: **1.413x**, reducing median E2E training-step time from **1.491 s to 1.055 s**.
- Once local FA is enabled, reducing the ring still adds **1.043x** over full-ring FA.
- Peak allocated memory for the combined path was roughly flat versus the full-ring Flex baseline: **23.65 GiB vs 23.18 GiB**.

### Independent repeat

A separate run with 3 warmup + 5 measured steps reproduced the ranking:

| Sliding path | Median step time | Median throughput | Speedup |
|---|---:|---:|---:|
| Full KV ring + Flex | 1.493 s | 21.94k tokens/s | 1.000x |
| Reduced KV ring + Flex | 1.180 s | 27.78k tokens/s | **1.266x** |
| Full KV ring + local FA | 1.111 s | 29.50k tokens/s | **1.344x** |
| Reduced KV ring + local FA | 1.063 s | 30.81k tokens/s | **1.404x** |

So, for this 32K Gemma4 E2B CP8 training configuration, the measured end-to-end speedup of the combined reduced-KV + local-FA path is consistently about **1.40-1.41x** over the full-ring Flex baseline.

**Review by HuiyingLi — APPROVED** [review]
Thank you so much!
